src/accountsdb/cache.zig

```diff
@@ -0,0 +1,331 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Slot = sig.core.time.Slot;
+const Pubkey = sig.core.pubkey.Pubkey;
+
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+
+/// Stores read-only in-memory copies of commonly used accounts
+pub const AccountsCache = struct {
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        // this account has since been mutated, and will not progress to newer slots
+        // TODO: when we start mutating accounts, make sure to set this field
+        is_dirty: std.atomic.Value(bool) = .{ .raw = false },
+        ref_count: std.atomic.Value(usize),
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) !CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = std.atomic.Value(usize).init(1),
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            const current_count = self.ref_count.load(.acquire);
+            if (current_count == 1) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            } else {
+                _ = self.ref_count.fetchSub(1, .acq_rel);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            _ = self.ref_count.fetchAdd(1, .acq_rel);
+            return self;
+        }
+    };
+
+    pub const LRU = LruCacheCustom(
+        .locking,
```

Comment:
Could you generally comment on the multi-threaded interaction with the AccountsCache itself, and each LRU entry, and/or any future plans regarding this?

Within AccountsDB the whole account cache is under a read-write lock, so I'm struggling to see when having each individual LRU entry be thread-safe in and of itself would be necessary.

Reply:
Down the line I'd like to get rid of the mutex wrapping the cache and make it internally locking.